### ISSUE TITLE
Webmin fails to use ECC certificates

### ISSUE THREAD

**Issue by wolfbeast** [webmin/webmin#1557]
We've been moving from RSA to ECC certificates and tried to use an ECC cert (wildcard) with Webmin SSL.
This failed spectacularly with a cipher mismatch. Apparently miniserv only accepts RSA-keyed certificates?

It would be great if ECC certs could be used.

**Comment by jcameron** [comment]
Webmin just uses the OpenSSL library to load and handle certs, so it should support ECC if the underlying library does.

How did you generate this cert? And what was the exact error message?

**Comment by wolfbeast** [comment]
I didn't run into it first hand because it was a colleague of mine installing it in webmin on a server under my domain. He stated the error was "cipher mismatch".
The certificate itself was generated with an OpenSSL CSR and signed by my CA (Sectigo) which is in live use on our websites (www.palemoon.org et al if you want to investigate the cert).

For ECC Certs the cipher suites are generally `TLS_ECDHE_ECDSA_*` and not `TLS_ECDHE_RSA_*` so that would likely be where the problem is. OpenSSL supports those suites, AFAIK.

**Comment by jcameron** [comment]
Which Linux distribution and version are you running Webmin on there? This could be an openssl version-specific issue..

**Comment by wolfbeast** [comment]
That would be CentOS 7.9.2009. A little dated but it shouldn't be that old in the OpenSSL department that something like this isn't supported, right?

**Comment by wolfbeast** [comment]
I just checked in detail, it has `OpenSSL 1.0.2k-fips` and with that I can successfully connect to an ECC-cert using server:

> New, TLSv1/SSLv3, Cipher is ECDHE-ECDSA-AES256-GCM-SHA384
Server public key is 384 bit
Secure Renegotiation IS supported

So I'd think that it should not be an OpenSSL problem per se on the O.S.

**Comment by wolfbeast** [comment]
It's been a few years and I ran into this problem again, being forced to use a self-signed cert on my webmin instance now as we've completely stopped using RSA for web certificates and the last usable one expired recently. This is not ideal, as I don't want to be forced to issue an RSA cert only for webmin. If there is a workaround for this, e.g. by making advanced miniserv conf entries or what not, I'm all ears.

**Comment by jcameron** [comment]
This should have been fixed a while ago! @wolfbeast can you tell us more detail about the error you're seeing now?

**Comment by wolfbeast** [comment]
Oh? Well I'm running into the same issue, still. if I install an ECC certificate, it fails with "no cypher overlap" like I originally reported...

**Comment by jcameron** [comment]
Are you getting this "Secure Connection Failed" error in your browser, or in Webmin when it tries to connect to another servr?

**Comment by wolfbeast** [comment]
This happens when I try to access webmin with my browser. Outbound connections aren't a problem as far as I know.
The only problem seems to be the internal web server used to serve webmin pages.

**Comment by jcameron** [comment]
Ok, and how did you generate and/or acquire this ECC cert?

**Comment by jcameron** [comment]
Thanks, I will try to reproduce this

**Comment by jcameron** [comment]
So I just tested this using an EC cert generated by Virtualmin, and I was able to use it in Webmin just fine.

What output do you get if you try the command `wget -O - https://localhost:10000` ?

**Comment by wolfbeast** [comment]
> So I just tested this using an EC cert generated by Virtualmin, and I was able to use it in Webmin just fine.
> What output do you get if you try the command `wget -O - https://localhost:10000` ?

```
[root@dev2 ~]# wget -O - https://localhost:10000
--2025-02-21 10:38:45--  https://localhost:10000/
Resolving localhost (localhost)... 127.0.0.1
Connecting to localhost (localhost)|127.0.0.1|:10000... connected.
OpenSSL: error:14077410:SSL routines:SSL23_GET_SERVER_HELLO:sslv3 alert handshake failure
Unable to establish SSL connection.
```
No idea how to interpret that.
For the record, connecting to nginx on the same host with the exact same certificate results in:
```
[root@dev2 ~]# wget -O - https://repo.palemoon.org
--2025-02-21 10:44:36--  https://repo.palemoon.org/
Resolving repo.palemoon.org (repo.palemoon.org)... 2a02:c206:2186:5496::1, 81.17.96.111
Connecting to repo.palemoon.org (repo.palemoon.org)|2a02:c206:2186:5496::1|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: unspecified [text/html]
Saving to: ‘STDOUT’
{content snipped}
```

**Comment by iliajie** [comment]
> [root@dev2 ~]# wget -O - https://localhost:10000
> --2025-02-21 10:38:45--  https://localhost:10000/
> Resolving localhost (localhost)... 127.0.0.1
> Connecting to localhost (localhost)|127.0.0.1|:10000... connected.
> OpenSSL: error:14077410:SSL routines:SSL23_GET_SERVER_HELLO:sslv3 alert handshake failure

This behavior is expected because Webmin doesn’t include the LE intermediate certificate by default in the chain. Browsers usually cache or fetch missing intermediates, so they work fine, but `wget` requires the full chain from the webserver.

Also, using “localhost” won’t work, since SSL certificates are issued for specific domain names.

In short, you should run instead:

```
wget --no-check-certificate -O - https://localhost:10000
```

**Comment by wolfbeast** [comment]
There's no difference.

(this is not an LE certificate, by the way, as I already explained above)
The intermediate cert is included in the certificate file (it includes the cert+intermediate), similar to what I would use with RSA which works.

The result is still:
```
[root@dev2 ~]# wget --no-check-certificate -O - https://localhost:10000
--2025-02-21 19:00:51--  https://localhost:10000/
Resolving localhost (localhost)... 127.0.0.1
Connecting to localhost (localhost)|127.0.0.1|:10000... connected.
OpenSSL: error:14077410:SSL routines:SSL23_GET_SERVER_HELLO:sslv3 alert handshake failure
Unable to establish SSL connection.
```

**Comment by wolfbeast** [comment]
For the record, here are the settings in use which work just fine with RSA or the self-signed one generated in webmin we currently have in use.

![Image](https://github.com/user-attachments/assets/d4e598bb-929a-4aff-ba97-418ce8026a28)

**Comment by iliajie** [comment]
What's the OS and version?

**Comment by wolfbeast** [comment]
CentOS Linux 7.9.2009 (I know it's EOL)

**Comment by iliajie** [comment]
Does the `openssl` command on CentOS 7 support ECC certificates?

**Comment by wolfbeast** [comment]
Yes, it does. Far as I know, CentOS enabled ECC in OpenSSL starting with version 6.5 -- all Centos 7 versions should support EC certificates. OpenSSL from the CLI shows various ec commands available, as well.

**Comment by wolfbeast** [comment]
Confirmed TLS1.2 and ECC are supported with openssl on the affected machine by connecting out from the CLI with `openssl s_client -connect` to an ECC server.

```
No client certificate CA names sent
Peer signing digest: SHA512
Server Temp Key: ECDH, P-256, 256 bits
---
SSL handshake has read 3450 bytes and written 415 bytes
---
New, TLSv1/SSLv3, Cipher is ECDHE-ECDSA-AES256-GCM-SHA384
Server public key is 256 bit
Secure Renegotiation IS supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
SSL-Session:
    Protocol  : TLSv1.2
    Cipher    : ECDHE-ECDSA-AES256-GCM-SHA384
    Session-ID: 43D4988F8C7435124C69697D64F2A97B00B62E2F4BA3992894C5E84D1E96AB1F
    Session-ID-ctx:
    Master-Key: 40032EAB28212EB7A5B9D8FC4F790FB124E175ED27C6E553C2E2B84DBCCD8CA9B15A50A52392A4BD18707A2D1D29651D
    Key-Arg   : None
    Krb5 Principal: None
    PSK identity: None
    PSK identity hint: None
    Start Time: 1740166994
    Timeout   : 300 (sec)
    Verify return code: 0 (ok)
---

**Comment by iliajie** [comment]
Rather give it a try like this:

```
openssl genpkey -algorithm ED25519 -out ed25519key.pem
openssl pkey -in ed25519key.pem -text
```

**Comment by wolfbeast** [comment]
I'm not sure what you're asking of me here. I'm not regenerating the key and getting a new cert. The cert is fine and works everywhere except in webmin and has been issued and installed already. My openssl connect test was in fact connecting to a server with that very same cert so that's not the issue. Additionally, we want EC, not EdDSA keys, anyway.
The issue seems to be specifically with miniserv.

**Comment by iliajie** [comment]
> I'm not sure what you're asking of me here.

If those commands work, then Ed25519 is supported on your system.

**Comment by wolfbeast** [comment]
Ed25519 isn't supported on the openssl version in use, *however*:
Like I said it shouldn't matter, since the cert is created with standard EC, not Ed25519 (on purpose, for backwards compat!). Ed25519 support shouldn't be necessary, so why are you asking me to check for support?

The cert uses:

> ANSI X9.62 elliptic curve prime256v1 (aka secp256r1, NIST P-256)

**Comment by iliajie** [comment]
Apologies, but what do you want from us? Are you insisting there is a bug or improvement we can make on our side? If so, does this issue occur on Rocky or Alma, version 8 or 9?

**Comment by wolfbeast** [comment]
... Excuse me? 
I thought it was clear. I'll repeat my initial issue:
Webmin doesn't accept a standard ECC certificate, while it does accept an RSA certificate issued by the same CA. That is a bug, IMO.
An algorithm that is not supported shouldn't influence acceptance of a certificate/key pair that doesn't use that algorithm. Or are you saying that miniserv won't accept any ECC certificates unless the underlying OpenSSL supports every possible algorithm? (Still a bug, in that case)

I have no easy way to test this on completely different OSes. The reason we're still using CentOS 7 is because of RH upending and destroying CentOS as a stable environment with v8 "Stream" being an unstable channel. I'm not familiar with "Rocky" or "Alma".

**Comment by wolfbeast** [comment]
> does this issue occur on Rocky or Alma, version 8 or 9?

We've migrated one of the systems to Almalinux 8 using ELevate now, and I can confirm this does not occur on AlmaLinux 8 -- the openssl version in use there is 1.1.1k-FIPS which does support ED25519 so probably satisfies what miniserv expects.

**Comment by jcameron** [comment]
I'll take another look at this with the prime256v1 key type you originally posted about..

**Comment by jcameron** [comment]
Ok I tested generating a self-signed prime256v1 EC key and it worked fine!

**Comment by jcameron** [comment]
I'm also the dev of miniserv, since it's part of Webmin :-) 

We're really just relying on the OpenSSL library and Net::SSLeay perl module to do the work of SSL negotiation though.

**Comment by wolfbeast** [comment]
Well, I don't really know what more information I can give here. OpenSSL 1.0.1k works from the CLI with the cert when used in NGinX, so maybe it's an issue in the perl module, then? If all miniserv does is offload to it, and the OpenSSL lib itself seems to be able to work just fine on its own with its binaries, that seems to be the broken link...

**Comment by jcameron** [comment]
Yeah, sorry I am not really able to debug this further as I can't re-produce it on my test systems.

Actually, which browser version are you running, and on which OS?
